### PR TITLE
Add new Lab Results endpoints

### PR DESCRIPTION
**Pull request by elliehastings**
## Add new Lab Results endpoints

* List Lab Results
* Add Lab Result Document
* List Changed Lab Results

Requests/responses have a subset of all possible options for our intended use cases but could be extended for full support in the future

## Breaking Changes

None

### CODE REVIEW

File: athenahealth/lab_results.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	Pagination *PaginationOptions
+}
+
+type listLabResultsResponse struct {
+	LabResults []*LabResult `json:"results"`
+
+	PaginationResponse
+}
+
+type ListLabResultsResult struct {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:49]: unfortunate stutter but understandable and would be the most reasonably expected name

File: athenahealth/lab_results.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+	Pagination *PaginationResult
+}
+
+// ListLabResults retrieves the patients laboratory results
+//
+// GET /v1/{practiceid}/chart/{patientid}/labresults
+//
+// https://docs.athenahealth.com/api/api-ref/lab-result#Get-patient's-lab-results
+func (h *HTTPClient) ListLabResults(ctx context.Context, patientID string, departmentID string, opts *ListLabResultsOptions) (*ListLabResultsResult, error) {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:60]: assuming that we're all on board with the proposal in this other PR https://github.com/eleanorhealth/go-athenahealth/pull/111, we should keep non-url path parameters that are required in the options struct as non-pointer fields.

File: athenahealth/lab_results.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	LabResultID  int    `json:"labresultid"`
+	Success      bool   `json:"success"`
+}
+
+// AddLabResultDocument creates a lab result document record of a specific patient
+//
+// POST /v1/{practiceid}/patients/{patientid}/documents/labresult
+//
+// https://docs.athenahealth.com/api/api-ref/document-type-lab-result#Add-lab-result-document-to-patient's-chart
+func (h *HTTPClient) AddLabResultDocumentReader(ctx context.Context, patientID string, departmentID string, opts *AddLabResultDocumentOptions) (int, error) {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:147]: same comment as above on arguments in options struct

File: athenahealth/lab_results.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	form := NewFormURLEncoder()
+
+	form.AddString("departmentid", departmentID)
+
+	if opts != nil {
+		if opts.AttachmentContents != nil {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:153]: we may want to surface an error if the attachment contents reader is `nil`

File: athenahealth/lab_results.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+			form.AddReader("attachmentcontents", newBase64Reader(opts.AttachmentContents))
+		}
+		if len(opts.AttachmentType) > 0 {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:156]: We may want to return an error if the attachment type is empty. Might be a good idea to capture any/all client-side errors and join. Checkout my comment here https://github.com/eleanorhealth/go-athenahealth/pull/111#discussion_r1293486940

File: athenahealth/lab_results.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+type ListChangedLabResultsOptions struct {
+	ShowPortalOnly             *bool     `json:"showportalonly"`
+	LeaveUnprocessed           *bool     `json:"leaveunprocessed"`
+	ShowProcessedEndDateTime   time.Time `json:"showprocessedenddatetime"`
+	ShowProcessedStartDateTime time.Time `json:"showprocessedstartdatetime"`
+
+	Pagination *PaginationOptions
+}
+
+type ListChangedLabResultsChangedLabResult struct {
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:208]: thoughts on just making this `ChangedLabResult`?

File: athenahealth/lab_results.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+	ProviderID           int    `json:"providerid"`
+	ReportStatus         string `json:"reportstatus"`
+	ResultStatus         string `json:"resultstatus"`
+	Status               string `json:"status"`
+	Subject              string `json:"subject"`
+	TieToOrderID         int    `json:"tietoorderid"`
+}
+
+type listChangedLabResultsResponse struct {
+	LabResults []*ListChangedLabResultsChangedLabResult `json:"labresults"`
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:251]: Do we want to call this key `ChangedLabResults` to make a distinction between the standard `LabResult` type?

File: athenahealth/lab_results.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+	if opts != nil {
+		if opts.ShowPortalOnly != nil {
+			q.Add("showportalonly", strconv.FormatBool(*opts.ShowPortalOnly))
+		}
+		if opts.LeaveUnprocessed != nil {
+			q.Add("leaveunprocessed", strconv.FormatBool(*opts.LeaveUnprocessed))
+		}
+		if !opts.ShowProcessedEndDateTime.IsZero() {
+			q.Add("ShowProcessedEndDateTime", opts.ShowProcessedEndDateTime.Format("01/02/2006 15:04:05"))
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:278]:
```suggestion
			q.Add("showprocessedenddatetime", opts.ShowProcessedEndDateTime.Format("01/02/2006 15:04:05"))
```

File: athenahealth/lab_results.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+		}
+		if !opts.ShowProcessedStartDateTime.IsZero() {
+			q.Add("ShowProcessedStartDateTime", opts.ShowProcessedStartDateTime.Format("01/02/2006 15:04:05"))
```
Comment (dillonstreator) [inline athenahealth/lab_results.go:281]:
```suggestion
			q.Add("showprocessedstartdatetime", opts.ShowProcessedStartDateTime.Format("01/02/2006 15:04:05"))
```

**Review by dillonstreator — COMMENTED** [review]
nice work ✨ 
left a few comments to address

File: athenahealth/lab_results.go (hunk 2 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:60]: Ah interesting, I missed that. Does make sense, and we should probably error if not set 👍

File: athenahealth/lab_results.go (hunk 4 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:153]: This gets into some divergence between how the athena API / client thinks about lab results and Eleanor does -- for us yes that should be an error case. For athena though it's seems valid to upload lab results with no attachments, for example if analytes and a result/interpretation are provided. Another one that might be good to chat more about.

File: athenahealth/lab_results.go (hunk 6 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:208]: How could you not love the name `ListChangedLabResultsChangedLabResult`? 😁 jk good call 👍

File: athenahealth/lab_results.go (hunk 7 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:251]: Yeah, I was thinking perhaps this should be consistent with athena's response but that makes more sense

File: athenahealth/lab_results.go (hunk 2 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:60]: We discussed live with the team and agreed we'd like to use a more enforced version of the current intended pattern of passing all required params as method arguments and optional params as an optional (pointer to) options struct, plus enforcing required field presence where it makes sense before making the call to athena. I've updated my branch to reflect this and @dillonstreator will update external parties / the README to reflect the decision and pattern.

File: athenahealth/lab_results.go (hunk 4 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:175]: Discussed and reviewed API and confirmed these are not required fields for this API so will not include error handling for those

File: athenahealth/lab_results.go (hunk 5 above)
Comment (elliehastings) [reply athenahealth/lab_results.go:178]: Addressed in above comments

File: athenahealth/lab_results.go (hunk 2 above)
Comment (dillonstreator) [reply athenahealth/lab_results.go:60]: https://github.com/eleanorhealth/go-athenahealth#method-signatures-required-vs-optional-fields

**Review by dillonstreator — APPROVED** [review]

**Pull request merged**